src/EnergyPlus/MixedAir.cc

```diff
@@ -299,7 +299,8 @@ void SimOASysComponents(EnergyPlusData &state, int const OASysNum, bool const Fi
                        OASysNum,
                        OAHeatCoil,
                        OACoolCoil,
-                       OAHX);
+                       OAHX,
+                       CompNum);
```

Comment:
These calls pass state.dataAirLoop->OutsideAirSys(OASysNum).ComponentIndex(CompNum) as an index to a component. This index = 0 on the first call. In GetOutsideAirSysInputs, the pointer based objects call a factory and save the object pointer in the compPointer list using this index. So for pointer based objects ComponentIndex and compPointer are already set when we get here. However, this argument is a reference so the Array1D models can pass back an index into this array. The pointer based objects are stomping on this index.

Reply:
This is confusing to explain. The state.dataAirLoop->OutsideAirSys(OASysNum).ComponentIndex(CompNum) value needs to be set in GetOutsideAirSysInputs where the index to the OA equipment list component is known. This index is not known in SimOAComponent because it is not passed, the array value is passed. So because of this the object pointer is saved in compNum = 0 if it were not set in GetOutsideAirSysInputs:

    state.dataAirLoop->OutsideAirSys(OASysNum).compPointer[compNum]->simulate()

compNum needs to be the index of the equipment in the OA System equipment list, not the equipment index.

In this loop, the loop counter CompNum needs to be saved in state.dataAirLoop->OutsideAirSys(OASysNum).ComponentIndex(CompNum) because that is the index where the factory pointer is saved (in GetOutsideAirSysInputs). So if the factory is called in SimOAComponent, the index to the pointer = 0 because it has not yet been set.

    for (int CompNum = 1; CompNum <= state.dataAirLoop->OutsideAirSys(OASysNum).NumComponents; ++CompNum) {
        CompType = state.dataAirLoop->OutsideAirSys(OASysNum).ComponentType(CompNum);
        CompName = state.dataAirLoop->OutsideAirSys(OASysNum).ComponentName(CompNum);
        SimOAComponent(state,
                       CompType,
                       CompName,
                       state.dataAirLoop->OutsideAirSys(OASysNum).ComponentTypeEnum(CompNum),
                       FirstHVACIteration,
                       state.dataAirLoop->OutsideAirSys(OASysNum).ComponentIndex(CompNum),

and when the factory is called, then the ComponentIndex should also be set:

    state.dataAirLoop->OutsideAirSys(OASysNum).ComponentIndex(CompNum) = CompNum;

which feels redundant but not doing this is causing the indexing issue.